Support clearing and updating state of individual mapped task instances

ephraimbuddy opened the pull request:
Provide support for UI to clear/update state of mapped task instances

TODO:
- [x] Add test at UI side

ephraimbuddy committed:
Allow marking/clearing mapped taskinstances from the UI

ephraimbuddy committed:
fixup! Allow marking/clearing mapped taskinstances from the UI

ephraimbuddy committed:
fixup! fixup! Allow marking/clearing mapped taskinstances from the UI

ephraimbuddy committed:
fixup! fixup! fixup! Allow marking/clearing mapped taskinstances from the UI

ephraimbuddy committed:
add tests

uranusjr commented on airflow/api/common/mark_tasks.py:
```diff
@@ -79,6 +79,7 @@ def _create_dagruns(
 def set_state(
     *,
     tasks: Iterable[Operator],
+    map_indexes: Optional[Iterable[int]] = None,
```
I thought about this a while ago, passing these as two arguments seems wrong because generally it only makes sense if either `tasks` are all non-mapped, or contain only one task. It makes very little sense to do something like

```python
set_state(tasks=[task_1, task_2], map_indexes=[1, 3])
```

Instead, we should probably make `tasks` take 2-tuples like this:

```python
def set_state(
    tasks: Iterable[Union[Operator, Tuple[MappedOperator, int]]],
    ...
): ...

set_state(
    tasks=[
        (task_1, 2),  # The specific mapped ti.
        task_2,  # The one ti if task_2 is not mapped, or all tis if mapped.
    ],
)
```

This would involve a larger change though, or at least that’s my impression when I last looked into this (before I got distracted by other issues).

ephraimbuddy replied:
Updated

ephraimbuddy committed:
fixup! fixup! fixup! fixup! Allow marking/clearing mapped taskinstances from the UI

ephraimbuddy committed:
fixup! fixup! fixup! fixup! fixup! Allow marking/clearing mapped taskinstances from the UI

ephraimbuddy committed:
fixup! fixup! fixup! fixup! fixup! fixup! Allow marking/clearing mapped taskinstances from the UI

jedcunningham commented on airflow/api/common/mark_tasks.py:
```diff
@@ -96,7 +96,8 @@ def set_state(
     on the schedule (but it will as for subdag dag runs if needed).
 
-    :param tasks: the iterable of tasks from which to work. task.task.dag needs to be set
+    :param tasks: the iterable of tasks or task, map_index tuple from which to work.
```
```suggestion
    :param tasks: the iterable of tasks or (task, map_index) tuples from which to work.
```

nit

jedcunningham commented on airflow/models/dag.py:
```diff
@@ -1789,6 +1817,8 @@ def clear(
         :param session: The sqlalchemy session to use
         :param dag_bag: The DagBag used to find the dags subdags (Optional)
         :param exclude_task_ids: A set of ``task_id`` that should not be cleared
+        :param exclude_task_ids_and_map_indexes: A set of ``task_id``,``map_index``
```
```suggestion
        :param exclude_task_ids_and_map_indexes: A set of ``(task_id, map_index)``
```

uranusjr commented on airflow/api/common/mark_tasks.py:
```diff
@@ -118,7 +119,9 @@ def set_state(
-    task_dags = {task.dag for task in tasks}
+    t_dags = {task.dag for task in tasks if not isinstance(task, tuple)}
+    t_dags_2 = {item[0].dag for item in tasks if isinstance(item, tuple)}
+    task_dags = t_dags | t_dags_2
```
```suggestion
    task_dags = {
        task[0].dag if isinstance(task, tuple) else task.dag
        for task in tasks
    }
```

uranusjr commented on airflow/models/dag.py:
```diff
@@ -1368,6 +1370,7 @@ def _get_task_instances(
         self,
         *,
         task_ids,
+        task_ids_and_map_indexes,
```
Maybe merge these like `set_states`? (same for other arguments below, including the `exclude` ones)

Also maybe add type annotations to these, it’s not easy to tell what should be passed into these at first sight.

uranusjr commented on airflow/models/dag.py:
```diff
@@ -1658,9 +1679,13 @@ def set_task_instance_state(
         task = self.get_task(task_id)
         task.dag = self
+        if not map_indexes:
+            map_indexes = [-1]
```
Intuitively I think not passing `map_idnexes` should default to setting states of all task instances of this task, not just -1 (matters if the task is mapped).

Another thing to consider is that, since this function is named `set_task_instance_state` (singular), it probably should only be able to set state to one task instance, i.e. it should only take one map index, not multiple. If we want to allow setting multiple mapped task instances, we should name the function `set_task_state` and deprecate this name.

uranusjr commented on airflow/www/views.py:
```diff
@@ -2025,6 +2028,14 @@ def clear(self):
         origin = get_safe_url(request.form.get('origin'))
         dag = current_app.dag_bag.get_dag(dag_id)
 
+        map_indexes = request.form.get('map_indexes')
```
There’s `getlist` available for this. To use it we need to format the argument list in the front end in a specific way, but that’s better than manually parsing the list, I think.

https://werkzeug.palletsprojects.com/en/2.1.x/datastructures/#werkzeug.datastructures.MultiDict.getlist

ephraimbuddy committed:
Apply suggestions from code review

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>
Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

ephraimbuddy committed:
fixup! Apply suggestions from code review

uranusjr committed:
Refactor to straighten up types

uranusjr committed:
Accept multiple map_index param from front end

This allows setting multiple instances of the same task to SUCCESS or
FAILED in one request. This is translated to multiple task specifier
tuples (task_id, map_index) when passed to set_state().

Also made some drive-through improvements adding types and clean some
formatting up.

uranusjr committed:
Introduce tuple_().in_() shim for MSSQL compat

bbovenzi reviewed (changes requested):
`/confirm`, `/failed` and `/success` work. But `/clear` is not processing multiple map_index params chained together.

The UI branch at `mapped-instance-actions` is up-to-date if you want to manually test it.

bbovenzi reviewed (approved):
Everything is working as expected from the UI now
